### PR TITLE
Fix real-time in Flow Controller

### PR DESCRIPTION
**Pull request by richiware**
## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Changes are ABI compatible. 
- *N/A* Changes are API compatible. 
- *N/A* New feature has been added to the `versions.md` file (if applicable).
- *N/A* New feature has been documented/Current behavior is correctly described in the documentation. 
    
- *N/A* Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by richiware** [comment]
@richiprosima please test linux

**Comment by richiware** [comment]
@richiware  please test mac

**Comment by richiware** [comment]
@richiprosima please test mac

**Comment by richiware** [comment]
@richiprosima please test this

**Comment by richiware** [comment]
@richiprosima please test this

**Comment by richiware** [comment]
@richiprosima please test this

**Comment by richiware** [comment]
@richiprosima please test this

**Comment by richiware** [comment]
@richiprosima please test this

**Comment by richiware** [comment]
@richiprosima please test this

**Comment by richiware** [comment]
@richiprosima please test mac

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
Partial review: headers and source. Pending review of tests.

File: include/fastdds/rtps/history/WriterHistory.h [hunk 1]
```diff
@@ -99,6 +112,10 @@ class WriterHistory : public rtps::History
     RTPS_DllAPI virtual bool remove_change_g(
             CacheChange_t* a_change);
 
+    RTPS_DllAPI virtual bool remove_change_g(
```
[the remark this answers is not in the document]
Comment (richiware) [reply include/fastdds/rtps/history/WriterHistory.h:115]: I dont' think so. Any  user's thread call that function.

File: src/cpp/fastdds/publisher/DataWriterHistory.cpp [hunk 2]
```diff
@@ -357,7 +364,7 @@ bool DataWriterHistory::remove_change_pub(
     std::lock_guard<RecursiveTimedMutex> guard(*this->mp_mutex);
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/fastdds/publisher/DataWriterHistory.cpp:373]: Fix in aaea68156

File: src/cpp/fastdds/publisher/DataWriterHistory.hpp [hunk 3]
```diff
@@ -174,8 +174,22 @@ class DataWriterHistory : public fastrtps::rtps::WriterHistory
     bool remove_change_pub(
             fastrtps::rtps::CacheChange_t* change);
 
-    virtual bool remove_change_g(
-            fastrtps::rtps::CacheChange_t* a_change);
+    /**
+     * Remove a change by the publisher History.
+     * @param change Pointer to the CacheChange_t.
+     * @param max_blocking_time
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/fastdds/publisher/DataWriterHistory.hpp]: Done in be2c0860b

File: src/cpp/rtps/history/History.cpp [hunk 4]
```diff
@@ -96,21 +104,27 @@ History::iterator History::remove_change_nts(
...
         CacheChange_t* ch)
+{
+    return History::remove_change(ch, std::chrono::steady_clock::now() + std::chrono::hours(24));
+}
+
+bool History::remove_change(
+        CacheChange_t* ch,
+        const std::chrono::time_point<std::chrono::steady_clock>& max_blocking_time)
 {
     std::lock_guard<RecursiveTimedMutex> guard(*mp_mutex);
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/rtps/history/History.cpp:123]: Done in 95907f12f

File: src/cpp/rtps/history/WriterHistory.cpp [hunk 5]
```diff
@@ -262,7 +288,7 @@ bool WriterHistory::remove_min_change()
     }
 
     std::lock_guard<RecursiveTimedMutex> guard(*mp_mutex);
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/rtps/history/WriterHistory.cpp:298]: Done in 95907f12f

File: src/cpp/rtps/history/ReaderHistory.cpp [hunk 6]
```diff
@@ -130,6 +130,15 @@ bool ReaderHistory::matches_change(
...
         bool release)
+{
+    return ReaderHistory::remove_change_nts(removal, std::chrono::steady_clock::now() + std::chrono::hours(24),
+                   release);
+}
+
+History::iterator ReaderHistory::remove_change_nts(
+        const_iterator removal,
+        const std::chrono::time_point<std::chrono::steady_clock>&,
+        bool release)
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/rtps/history/ReaderHistory.cpp]: Done in db1bca5ec

File: src/cpp/rtps/history/History.cpp [hunk 7]
```diff
@@ -71,6 +71,14 @@ bool History::matches_change(
...
         const_iterator removal,
         bool release)
+{
+    return History::remove_change_nts(removal, std::chrono::steady_clock::now() + std::chrono::hours(24), release);
+}
+
+History::iterator History::remove_change_nts(
+        const_iterator removal,
+        const std::chrono::time_point<std::chrono::steady_clock>&,
+        bool release)
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/rtps/history/History.cpp]: Done in db1bca5ec

File: src/cpp/fastdds/subscriber/history/DataReaderHistory.cpp [hunk 8]
```diff
@@ -664,6 +664,15 @@ void DataReaderHistory::check_and_remove_instance(
...
         bool release)
+{
+    return DataReaderHistory::remove_change_nts(removal, std::chrono::steady_clock::now() + std::chrono::hours(
+                       24), release);
+}
+
+ReaderHistory::iterator DataReaderHistory::remove_change_nts(
+        ReaderHistory::const_iterator removal,
+        const std::chrono::time_point<std::chrono::steady_clock>&,
+        bool release)
```
[the remark this answers is not in the document]
Comment (richiware) [reply src/cpp/fastdds/subscriber/history/DataReaderHistory.cpp]: Done in db1bca5ec

**Comment by richiware** [comment]
@richiprosima please test this

**Review by MiguelCompany — APPROVED** [review]
LGTM with green CI

**Pull request merged**